src/Microsoft.Extensions.FileProviders.Physical/CachingFileSystemProvider.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.FileProviders.Physical
+{
+    public class CachingFileProvider: IFileProvider, IDisposable
+    {
+        private readonly IFileProvider _baseProvider;
+        private readonly ConcurrentLruCache<string, IFileInfo> _fileCache;
+        private readonly ConcurrentLruCache<string, IDirectoryContents> _directoryCache;
+        private Func<string, IFileInfo> _getFileInfo;
+        private Func<string, IDirectoryContents> _getDirectoryContents;
+        private IDisposable _changeCallbackRegistration;
+
+        public CachingFileProvider(IFileProvider baseProvider, int fileCapacity, int directoryCapacity, string watchFilter)
+        {
+            _baseProvider = baseProvider;
+            _fileCache = new ConcurrentLruCache<string, IFileInfo>(fileCapacity);
+            _directoryCache = new ConcurrentLruCache<string, IDirectoryContents>(directoryCapacity);
+            if (watchFilter != null)
+            {
+                _changeCallbackRegistration = _baseProvider.Watch(watchFilter).RegisterChangeCallback(Invalidate, null);
```

Comment:
We are trying to decrease amount of `GetFileAttributesEx` calls `PhysicalFileProvider` does and improve `StaticFiles` performance. User that adds this provider should know why is he doing it, Also passing `null` to `watchFilter` disables watcher.